File: backend/app/schemas/application.py
import uuid
from datetime import date, datetime

from pydantic import BaseModel, Field

APPLICATION_STATUS_VALUES = frozenset(
    {
        "Saved",
        "Applied",
        "OA",
        "Interview",
        "Final Round",
        "Offer",
        "Rejected",
        "Ghosted",
    }
)


class ApplicationCreate(BaseModel):
    employer_id: uuid.UUID
    role: str = Field(min_length=1, max_length=500)
    salary: str | None = Field(default=None, max_length=255)
    work_mode: str | None = Field(default=None, max_length=32)
    source_url: str | None = Field(default=None, max_length=2000)
    status: str = "Saved"
    applied_at: date | None = None
    deadline: date | None = None
    next_action_at: datetime | None = None
    notes: str = ""
    interview_feedback: str | None = None
    recruiter_interactions: str | None = None
    improvement_points: str | None = None


class ApplicationUpdate(BaseModel):
    employer_id: uuid.UUID | None = None
    role: str | None = Field(default=None, min_length=1, max_length=500)
    salary: str | None = Field(default=None, max_length=255)
    work_mode: str | None = Field(default=None, max_length=32)
    source_url: str | None = Field(default=None, max_length=2000)
    status: str | None = None
    applied_at: date | None = None
    deadline: date | None = None
    next_action_at: datetime | None = None
    notes: str | None = None
    resume_document_id: uuid.UUID | None = None
    interview_feedback: str | None = None
    recruiter_interactions: str | None = None
    improvement_points: str | None = None


class ApplicationRead(BaseModel):
    id: uuid.UUID
    employer_id: uuid.UUID
    role: str
    salary: str | None
    work_mode: str | None
    source_url: str | None
    status: str
    applied_at: date | None
    deadline: date | None
    next_action_at: datetime | None
    notes: str
    resume_document_id: uuid.UUID | None
    document_ids: list[uuid.UUID]
    interview_feedback: str | None
    recruiter_interactions: str | None
    improvement_points: str | None
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}

File: src/domain/application.ts
import type { ApplicationStatus } from "./applicationStatus";
import type { Employer } from "./employer";
import { employerName } from "./employer";
import type { WorkMode } from "./workMode";

export interface Application {
  id: string;
  employerId: string;
  role: string;
  salary: string | null;
  workMode: WorkMode | null;
  sourceUrl: string | null;
  status: ApplicationStatus;
  appliedAt: string | null;
  deadline: string | null;
  nextActionAt: string | null;
  notes: string;
  /** Primary resume used for this application (document id). */
  resumeDocumentId: string | null;
  /** All documents attached to this application (resume, cover letter, etc.). */
  documentIds: string[];
  interviewFeedback?: string | null;
  recruiterInteractions?: string | null;
  improvementPoints?: string | null;
}

let idCounter = 0;

function nextId(): string {
  idCounter += 1;
  return `app-${idCounter}`;
}

export function createApplicationDraft(input: {
  employerId: string;
  role: string;
}): Application {
  return {
    id: nextId(),
    employerId: input.employerId,
    role: input.role,
    salary: null,
    workMode: null,
    sourceUrl: null,
    status: "Saved",
    appliedAt: null,
    deadline: null,
    nextActionAt: null,
    notes: "",
    resumeDocumentId: null,
    documentIds: [],
  };
}

export function filterApplicationsByStatus(
  rows: Application[],
  status: ApplicationStatus,
): Application[] {
  return rows.filter((r) => r.status === status);
}

export function sortApplicationsByEmployerName(
  rows: Application[],
  employers: Employer[],
): Application[] {
  return [...rows].sort((a, b) =>
    employerName(employers, a.employerId).localeCompare(
      employerName(employers, b.employerId),
      undefined,
      { sensitivity: "base" },
    ),
  );
}
